File: datasets/C++/Chatgpt/DataStructures/easy5.cpp
#include <iostream>
#include <map>
#include <queue>
#include <unordered_map>
#include <unordered_set>
using namespace std;

struct Node {
    int data;
    Node *left, *right;
    Node(int d) : data(d), left(NULL), right(NULL) {}
};

void topView(Node* root) {
    if (!root) return;
    map<int, int> hd_map;
    queue<pair<Node*, int>> q;
    q.push({root, 0});
    while (!q.empty()) {
        auto [node, hd] = q.front(); q.pop();
        if (!hd_map.count(hd))
            hd_map[hd] = node->data;
        if (node->left) q.push({node->left, hd - 1});
        if (node->right) q.push({node->right, hd + 1});
    }
    for (auto &p : hd_map) cout << p.second << " ";
    cout << endl;
}

int main() {
    int n;
    cout << "Enter number of nodes:\n";
    cin >> n;
    if (n == 0) return 0;

    unordered_map<int, Node*> nodes;
    cout << "Enter node values (unique IDs):\n";
    for (int i = 0; i < n; i++) {
        int val; cin >> val;
        nodes[val] = new Node(val);
    }

    int e;
    cout << "Enter number of edges:\n";
    cin >> e;
    if (e > n - 1) {
        cout << "Error: too many edges\n";
        return 0;
    }

    unordered_set<int> children;
    cout << "Enter edges in format: parent child L/R\n";
    for (int i = 0; i < e; i++) {
        int p, c; char dir;
        cin >> p >> c >> dir;
        if (!nodes.count(p) || !nodes.count(c)) {
            cout << "Error: invalid node ID\n";
            return 0;
        }
        children.insert(c);
        if (dir == 'L') nodes[p]->left = nodes[c];
        else if (dir == 'R') nodes[p]->right = nodes[c];
        else {
            cout << "Error: direction must be L or R\n";
            return 0;
        }
    }

    // Find root (node that is never a child)
    int rootVal = -1;
    for (auto &kv : nodes) {
        if (children.find(kv.first) == children.end()) {
            rootVal = kv.first;
            break;
        }
    }
    if (rootVal == -1) {
        cout << "Error: no root found\n";
        return 0;
    }

    topView(nodes[rootVal]);
}

File: datasets/Python/Chatgpt/DataStructures/easy5.py
from collections import deque

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def build_tree_manual():
    n = int(input("Enter number of nodes: "))
    if n == 0:
        return None
    print("Enter node values (unique IDs):")
    node_vals = list(map(int, input().split()))
    if len(node_vals) != n:
        print("Error: number of values doesn't match node count")
        return None
    nodes = {val: Node(val) for val in node_vals}

    e = int(input("Enter number of edges: "))
    if e > n - 1:
        print("Error: too many edges")
        return None
    print("Enter edges in format: parent child L/R")
    for _ in range(e):
        p, c, d = input().split()
        p, c = int(p), int(c)
        if p not in nodes or c not in nodes:
            print("Error: invalid node ID")
            return None
        if d.upper() == 'L':
            nodes[p].left = nodes[c]
        elif d.upper() == 'R':
            nodes[p].right = nodes[c]
        else:
            print("Error: direction must be L or R")
            return None
    return nodes[node_vals[0]]

def topView(root):
    if not root:
        return
    q = deque([(root, 0)])
    hd_map = {}
    while q:
        node, hd = q.popleft()
        if hd not in hd_map:
            hd_map[hd] = node.data
        if node.left:
            q.append((node.left, hd - 1))
        if node.right:
            q.append((node.right, hd + 1))
    for hd in sorted(hd_map):
        print(hd_map[hd], end=' ')
    print()

# ---- TESTING ----
if __name__ == "__main__":
    root = build_tree_manual()
    topView(root)